language/vm/vm_runtime/vm_runtime_types/src/native_structs/vector.rs

```diff
@@ -58,84 +64,123 @@ impl NativeVector {
+        }) {
+            Ok(_) => NativeReturnStatus::Success {
                 cost: PUSH_BACK_COST,
                 return_values: vec![],
-            })
-            .unwrap_or(NativeReturnStatus::InvalidArguments)
+            },
+            Err(err) => NativeReturnStatus::InvariantError(err),
+        }
     }
 
     pub fn native_borrow(mut args: VecDeque<Value>) -> NativeReturnStatus {
         if args.len() != 2 {
-            return NativeReturnStatus::InvalidArguments;
+            let msg = format!(
+                "wrong number of arguments for borrow expected 2 found {}",
+                args.len()
+            );
+            return NativeReturnStatus::InvariantError(
+                VMStatus::new(StatusCode::UNREACHABLE).with_message(msg),
+            );
         }
         let reference = get_vector_ref!(args);
         let idx = pop_arg!(args, u64);
-        match reference.get_native_struct_reference(|native_val| {
-            get_vector(native_val)?
-                .0
-                .get(idx as usize)
-                .map(MutVal::clone)
+        match reference.get_native_struct_reference(|struct_ref| {
+            get_vector(struct_ref).and_then(|native_vec| match native_vec.0.get(idx as usize) {
+                Some(val) => Ok(val.clone()),
+                None => Err(VMStatus::new(StatusCode::NATIVE_FUNCTION_ERROR)
+                    .with_sub_status(INDEX_OUT_OF_BOUNDS)),
+            })
         }) {
-            Some(v) => NativeReturnStatus::Success {
+            Ok(v) => NativeReturnStatus::Success {
                 cost: BORROW_COST,
                 return_values: vec![v],
             },
-            None => NativeReturnStatus::Aborted {
-                cost: BORROW_COST,
-                error_code: INDEX_OUT_OF_BOUNDS,
+            Err(err) => match err.major_status {
+                StatusCode::NATIVE_FUNCTION_ERROR => NativeReturnStatus::Aborted {
+                    cost: BORROW_COST,
+                    error_code: err,
+                },
+                _ => NativeReturnStatus::InvariantError(err),
             },
         }
     }
 
     pub fn native_pop(mut args: VecDeque<Value>) -> NativeReturnStatus {
         if args.len() != 1 {
-            return NativeReturnStatus::InvalidArguments;
+            let msg = format!(
+                "wrong number of arguments for pop expected 1 found {}",
+                args.len()
+            );
+            return NativeReturnStatus::InvariantError(
+                VMStatus::new(StatusCode::UNREACHABLE).with_message(msg),
+            );
         }
 
         let reference = get_vector_ref!(args);
-        match reference.mutate_native_struct(|native_val| {
-            get_mut_vector(native_val)?.0.pop().map(MutVal::into_value)
+        match reference.mutate_native_struct(|struct_ref| {
+            get_mut_vector(struct_ref).and_then(|native_vec| match native_vec.0.pop() {
+                Some(val) => val.into_value(),
```

Comment:
Can use `ok_or_else` to get rid of the match here. I don't have a strong opinion tho..

Reply:
not sure how to do that but I am always a bit vague on some of those rust-ism.
My understanding of ok_or_else() is that Some(val) => Ok(val) and None => error returned by the closure.
Here we need to compute something on the result of Some(val) thus I put the match, thoughts?